x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/http/StreamingHttpResultPublisher.java

```diff
@@ -0,0 +1,244 @@
+
+/**
+ * <p>Streams responses from Apache's HttpAsyncResponseConsumer to Java's Flow.Publisher.</p>
+ *
+ * <p>The ActionListener is called once when the HttpResponse is initially received to establish the Flow.  All subsequent responses and
+ * errors will be sent through the Flow API.  Consumers are expected to call #onSubscribe with a
+ * {@link java.util.concurrent.Flow.Subscriber} when the ActionListener is called.  Consumers can then begin using the resulting
+ * {@link java.util.concurrent.Flow.Subscription} to request HttpResults from this publisher.</p>
+ *
+ * <p>Consumers should expect the first HttpResult to represent the overall HttpResponse.  Apache opens the channel with an HttpResponse
+ * before it starts sending response bytes.  If the HttpResponse is an error, Apache may only send an HttpResponse with an HttpEntity,
+ * so this publisher will send a single HttpResult. If the HttpResponse is healthy, Apache will send an HttpResponse with or without
+ * the HttpEntity.</p>
+ */
+class StreamingHttpResultPublisher implements HttpAsyncResponseConsumer<HttpResponse>, Flow.Publisher<HttpResult> {
+    private final HttpSettings settings;
+    private final ActionListener<Flow.Publisher<HttpResult>> listener;
+
+    // used to manage the HTTP response
+    private volatile HttpResponse response;
+    private volatile Exception ex;
+
+    // used to control the state of this publisher (Apache) and its interaction with its subscriber
+    private final Deque<HttpResult> resultQueue = new ConcurrentLinkedDeque<>();
+    private final PendingRequestCount requestCount = new PendingRequestCount();
+    private final AtomicBoolean isDone = new AtomicBoolean(false);
+    private final AtomicBoolean subscriptionCanceled = new AtomicBoolean(false);
+    private volatile Flow.Subscriber<? super HttpResult> subscriber;
+
+    private final RequestBasedTaskRunner taskRunner;
+
+    // used to control the flow of data from the Apache client, if we're producing more bytes than we can consume then we'll pause
+    private final AtomicLong totalByteSize = new AtomicLong(0);
+    private final Object ioLock = new Object();
+    private volatile IOControl savedIoControl;
+
+    StreamingHttpResultPublisher(ThreadPool threadPool, HttpSettings settings, ActionListener<Flow.Publisher<HttpResult>> listener) {
+        this.settings = Objects.requireNonNull(settings);
+        this.listener = Objects.requireNonNull(listener);
+
+        this.taskRunner = new RequestBasedTaskRunner(new OffloadThread(), threadPool, UTILITY_THREAD_POOL_NAME);
+    }
+
+    @Override
+    public void responseReceived(HttpResponse httpResponse) throws IOException {
+        this.response = httpResponse;
+        this.resultQueue.offer(HttpResult.create(settings.getMaxResponseSize(), response));
+        this.listener.onResponse(this);
+    }
+
+    @Override
+    public void subscribe(Flow.Subscriber<? super HttpResult> subscriber) {
+        this.subscriber = subscriber;
```

Comment:
I think we can only have 1 subscriber, a second would overwrite the first. What's the best thing to do here? should any further subscribers be failed after the first?

Reply:
Yeah good catch - the spec says we should either return an IllegalStateException or have a queue of Subscribers, I think I'm going with the former until we need the latter?